src/gradualizer_lib.erl

```diff
@@ -90,58 +91,114 @@ reverse_graph(G) ->
-pick_value(?type(atom)) ->
+pick_value(?type(atom), _TEnv) ->
     {atom, erl_anno:new(0), a};
-pick_value({atom, _, A}) ->
+pick_value({atom, _, A}, _TEnv) ->
     {atom, erl_anno:new(0), A};
-pick_value({ann_type, _, [_, Ty]}) ->
-    pick_value(Ty);
-pick_value(?type(union, [Ty|_])) ->
-    pick_value(Ty);
-pick_value(?type(tuple, any)) ->
+pick_value({ann_type, _, [_, Ty]}, TEnv) ->
+    pick_value(Ty, TEnv);
+pick_value(?type(union, [Ty|_]), TEnv) ->
+    pick_value(Ty, TEnv);
+pick_value(?type(tuple, any), _TEnv) ->
     {tuple, erl_anno:new(0), []};
-pick_value(?type(tuple, Tys)) ->
-    {tuple, erl_anno:new(0), [pick_value(Ty) || Ty <- Tys]};
-pick_value(?type(record, [{atom, _, RecordName}])) ->
+pick_value(?type(tuple, Tys), TEnv) ->
+    {tuple, erl_anno:new(0), [pick_value(Ty, TEnv) || Ty <- Tys]};
+pick_value(?type(record, [{atom, _, RecordName}]), _TEnv) ->
     {record, erl_anno:new(0), RecordName, []};
-pick_value(?type(record, [{atom, _, RecordName} | Tys])) ->
+pick_value(?type(record, [{atom, _, RecordName} | Tys]), TEnv) ->
     MFields = [
-        {record_field, erl_anno:new(0), {atom, erl_anno:new(0), FieldName}, pick_value(Ty)}
+        {record_field, erl_anno:new(0), {atom, erl_anno:new(0), FieldName}, pick_value(Ty, TEnv)}
         || ?type(field_type, [{atom, _, FieldName}, Ty]) <- Tys
     ],
     {record, erl_anno:new(0), RecordName, MFields};
-pick_value(?type(list)) ->
+pick_value(?type(list), _TEnv) ->
     {nil, erl_anno:new(0)};
-pick_value(?type(list,_)) ->
+pick_value(?type(list,_), _TEnv) ->
     {nil, erl_anno:new(0)};
-pick_value(?type(nil)) ->
+pick_value(?type(nil), _TEnv) ->
     {nil, erl_anno:new(0)};
 %% The ?type(range) is a different case because the type range
 %% ..information is not encoded as an abstract_type()
 %% i.e. {type, Anno, range, [{integer, Anno2, Low}, {integer, Anno3, High}]}
-pick_value(?type(range, [{_TagLo, _, neg_inf}, Hi = {_TagHi, _, _Hi}])) ->
-    %% pick_value(Hi);
+pick_value(?type(range, [{_TagLo, _, neg_inf}, Hi = {_TagHi, _, _Hi}]), _TEnv) ->
+    %% pick_value(Hi, TEnv);
     Hi;
-pick_value(?type(range, [Lo = {_TagLo, _, _Lo}, {_TagHi, _, _Hi}])) ->
-    %% pick_value(Lo).
-    Lo.
+pick_value(?type(range, [Lo = {_TagLo, _, _Lo}, {_TagHi, _, _Hi}]), _TEnv) ->
+    %% pick_value(Lo, TEnv).
+    Lo;
+pick_value(Type, TEnv)
+  when element(1, Type) =:= remote_type; element(1, Type) =:= user_type ->
+    {Kind, Anno, Name, Args} = case Type of
+                                   {remote_type, Anno, [_, {atom, _, Name}, Args]} ->
+                                       {remote_type, Anno, Name, Args};
+                                   {user_type, Anno, Name, Args} ->
+                                       {user_type, Anno, Name, Args}
+                               end,
+    case get_type_definition(Type, TEnv, [annotate_user_types]) of
+        {ok, Ty} ->
+            pick_value(Ty, TEnv);
+        opaque ->
+            {var, Anno, '_Opaque'};
```

Comment:
If there are multiple opaque types in a picked value, using the same variable name would imply all opaque values are equal, as in `{picked_tuple, _Opaque, _Opaque}`. That's why I suggested adding a unique integer to the name or use just a `_`. Not a big deal though since this only for an error message. There is a point in not flooding the atom table as well.

Reply:
Indeed, this might be slightly misleading if more than one opaque type is used... I wanted to avoid passing extra mappings from types to names, but it shouldn't be hard to add - I'll do it in this PR.

Reply:
Is there a need for passing around extra mappings? I was thinking just

```Erlang
list_to_atom("Opaque" ++ integer_to_list(erlang:unique_integer([positive])))
```

Numbers might look a bit random in this way but at least they would be unique.

Reply:
> passing extra mappings from types to names

The same type used twice doesn't mean it's the same value, so I think it would actually be wrong to use the same variable twice in this case.